src/bin/pg_autoctl/pgsql.c

```diff
@@ -654,43 +721,265 @@ pgsql_drop_replication_slots(PGSQL *pgsql)
+												  "), \n"
+												  "dropped as ("
+												  " SELECT slot_name, pg_drop_replication_slot(slot_name) "
+												  "   FROM pg_replication_slots pgrs LEFT JOIN nodes USING(slot_name) "
+												  "  WHERE nodes.slot_name IS NULL "
+												  ") \n"
+												  "SELECT 'drop', slot_name, NULL::pg_lsn FROM dropped";
+
+	Oid paramTypes[NODE_ARRAY_MAX_COUNT * 2] = { 0 };
+	const char *paramValues[NODE_ARRAY_MAX_COUNT * 2] = { 0 };
+	ReplicationSlotMaintainContext context = { 0 };
+
+	int paramCount = BuildNodesArrayValues(nodeArray,
+										   paramTypes, (char **) paramValues,
+										   values, BUFSIZE);
+
+	/* add the computed ($1,$2), ... string to the query "template" */
+	bytes = snprintf(sql, 2 * BUFSIZE, sqlTemplate, values);
+
+	if (bytes > 2 * BUFSIZE)
+	{
+		/* shouldn't happen because we only support up to 12 nodes */
+		log_error("Failed to prepare the SQL query for "
+				  "pgsql_replication_slot_maintain");
+		return false;
+	}
+
+	return pgsql_execute_with_params(pgsql, sql,
+									 paramCount, paramTypes, paramValues,
+									 &context,
+									 parseReplicationSlotMaintain);
 }
 
 
 /*
- * pgsql_set_synchronous_standby_names set synchronous_standby_names ton the
- * local Postgres to the value computed on the pg_auto_failover monitor.
+ * pgsql_replication_slot_maintain maintains replication slots that belong to
+ * other standby nodes. We call that function on the standby nodes, where the
+ * slots are maintained manually just in case we need them at failover.
  */
 bool
-pgsql_set_synchronous_standby_names(PGSQL *pgsql,
-									char *synchronous_standby_names)
+pgsql_replication_slot_maintain(PGSQL *pgsql, NodeAddressArray *nodeArray)
 {
-	char quoted[BUFSIZE] = { 0 };
-	GUC setting = { "synchronous_standby_names", quoted };
+	int bytes;
+	char sql[2 * BUFSIZE] = { 0 };
+	char values[BUFSIZE] = { 0 };
+	char *sqlTemplate =
+		"WITH nodes(slot_name, lsn) as ("
+		" SELECT '" REPLICATION_SLOT_NAME_DEFAULT "_' || id, lsn"
+												  "   FROM (%s) as sb(id, lsn) "
+												  "), \n"
+												  "dropped as ("
+												  " SELECT slot_name, pg_drop_replication_slot(slot_name) "
+												  "   FROM pg_replication_slots pgrs LEFT JOIN nodes USING(slot_name) "
+												  "  WHERE nodes.slot_name IS NULL "
+												  "), \n"
+												  "advanced as ("
+												  "SELECT a.slot_name, a.end_lsn"
+												  "  FROM pg_replication_slots JOIN nodes USING(slot_name), "
+												  "       LATERAL pg_replication_slot_advance(slot_name, lsn) a"
+												  " WHERE nodes.lsn <> '0/0'"
+												  "), \n"
+												  "created as ("
+												  "SELECT c.slot_name, c.lsn "
+												  "  FROM nodes LEFT JOIN pg_replication_slots pgrs USING(slot_name), "
+												  "       LATERAL pg_create_physical_replication_slot(slot_name, true) c"
+												  " WHERE pgrs.slot_name IS NULL "
+												  ") \n"
```

Comment:
I think it makes sense to make these CTEs all look as similar as possible. That way the differences can be seen quickly and it makes it clear that they are all doing a very similar thing.

1. You should be able to remove the LATERAL join by doing either:
```sql
SELECT slot_name, (pg_replication_slot_advance(slot_name, lsn)).end_lsn ...
```
or 
```sql
SELECT (pg_replication_slot_advance(slot_name, lsn)).* ...
```
I'm fine with either of these.



2. Let's keep the tables in the same order in `created` and make it a RIGHT JOIN.

Reply:
The notation `(funcall()).*` has a better avoided side-effect; it calls the function as many times as there are parameters. LATERAL JOIN is wonderful because of that, much better than using the SELECT projection...

The main problem I have with that code now, after the fact, is that in case we fail to advance, create, or drop a single replication slot, then we stall progress for all the others. I think I might go on and have an explicit loop for create or advance and then a specific call for drop ; where we don't have a list to walk anyway...